core/peerset/src/peersstate.rs

```diff
@@ -187,21 +191,145 @@ impl PeersState {
 				}
 				Some(to_try)
 			})
-			.map(|(peer_id, state)| (peer_id.clone(), state));
+			.map(|(peer_id, _)| peer_id.clone());
 
-		if let Some((peer_id, state)) = outcome {
+		if let Some(peer_id) = outcome {
 			Some(NotConnectedPeer {
-				state,
+				state: self,
 				peer_id: Cow::Owned(peer_id),
-				num_in: &mut self.num_in,
-				num_out: &mut self.num_out,
-				max_in: self.max_in,
-				max_out: self.max_out,
 			})
 		} else {
 			None
 		}
 	}
+
+	fn disconnect(&mut self, peer_id: &PeerId) {
+		let is_priority = self.is_priority(peer_id);
+		if let Some(mut node) = self.nodes.get_mut(peer_id) {
```

Comment:
That's defensive programming again :-/
Instead of detecting bugs we just ignore them :(